guest-js/index.ts

```diff
@@ -1,114 +1,128 @@
-import { invoke } from '@tauri-apps/api/core'
+import { invoke } from '@tauri-apps/api/core';
+import { listen, UnlistenFn } from '@tauri-apps/api/event';
 
-/**
- * Creates a download operation.
- *
- * @param key - The key identifier.
- * @param url - The download URL  for the resource.
- * @param path - The download path on the filesystem.
- * @returns - The download operation.
- */
-export async function create(key: string, url: string, path: string): Promise<Download> {
-  return await invoke<Download>('plugin:download|create', {
-    key, url, path,
-  });
-}
+export class Download implements DownloadRecord {
+   public key: string;
+   public url: string;
+   public path: string;
+   public progress: number;
+   public state: DownloadState;
 
-/**
- * Gets all download operations.
- *
- * @returns - The list of download operations.
- */
-export async function list(): Promise<Download[]> {
-  return await invoke<Download[]>('plugin:download|list');
+   public constructor(record: DownloadRecord) {
+      this.key = record.key;
+      this.url = record.url;
+      this.path = record.path;
+      this.progress = record.progress;
+      this.state = record.state;
+   }
+
+   /**
+   * Starts the download.
+   * @returns A promise with the updated download.
+   */
+   public async start(): Promise<Download> {
+      return new Download(await invoke('plugin:download|start', { key: this.key }));
+   }
+
+   /**
+   * Cancels the download.
+   * @returns A promise with the updated download.
+   */
+   public async cancel(): Promise<Download> {
+      return new Download(await invoke('plugin:download|cancel', { key: this.key }));
+   }
+
+   /**
+   * Pauses the download.
+   * @returns A promise with the updated download.
+   */
+   public async pause(): Promise<Download> {
+      return new Download(await invoke('plugin:download|pause', { key: this.key }));
+   }
+
+   /**
+   * Resumes the download.
+   * @returns A promise with the updated download.
+   */
+   public async resume(): Promise<Download> {
+      return new Download(await invoke('plugin:download|resume', { key: this.key }));
+   }
+
+   /**
+   * Listen for changes to the download.
```

Comment:
Should we add a warning to this doc block that makes it clear that if you call this `listen` method you _must_ call the returned `unlisten` function when done (e.g. when unmounting a component) to avoid memory leaks? As-is, it may be reasonable to think that `unlisten` is just for when you want to stop reacting to events, not an essential step to avoid memory leaks.

Reply:
Good idea. Done.